### PR TITLE
core: retry support part 1 followup, drainedSubstreams should not contain closed ones

### PR DESCRIPTION
**Pull request by dapengzhang0**
2 commits:
- the first one is pure refactoring to add Substream class,
- the second is not to add closed substreams to drainedSubstreams.

### CODE REVIEW

**Comment by dapengzhang0** [comment]
cc @ericgribkoff

File: core/src/main/java/io/grpc/internal/RetriableStream.java [hunk 1]
```diff
@@ -536,6 +548,21 @@ State drained(Substream substream) {
       return new State(buffer, drainedSubstreams, winningSubstream, cancelled, passThrough);
     }
 
+    /** The given substream is closed. */
+    @CheckReturnValue
+    @GuardedBy("lock")
+    State closed(Substream substream) {
+      substream.closed = true;
+      if (this.drainedSubstreams.contains(substream)) {
+        Set<Substream> drainedSubstreams = new HashSet<Substream>();
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/RetriableStream.java]: Use the `HashSet(Collection)` copy constructor.

File: core/src/test/java/io/grpc/internal/RetriableStreamTest.java [hunk 2]
```diff
@@ -683,6 +688,48 @@ public void messageAvailable() {
...
+    final ClientStream mockStream3 = mock(ClientStream.class);
+    final ClientStream mockStream2 =
+        mock(
+            ClientStream.class,
+            delegatesTo(
+                new NoopClientStream() {
+                  @Override
+                  public void start(ClientStreamListener listener) {
+                    // closed while draning and retry
+                    doReturn(true).when(retriableStreamRecorder).shouldRetry();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/RetriableStreamTest.java]: Changing stub return values within this start seems weird, as I'm not sure why they are related. It also makes the test harder to follow. Why isn't this just done early in `closedWhileDraining()`? It doesn't look like it ever will return `false` now.

File: core/src/test/java/io/grpc/internal/RetriableStreamTest.java [hunk 3]
```diff
@@ -683,6 +688,48 @@ public void messageAvailable() {
     verify(masterListener).messagesAvailable(messageProducer);
   }
 
+  @Test
+  public void closedWhileDraining() {
+    ClientStream mockStream1 = mock(ClientStream.class);
+    doReturn(mockStream1).when(retriableStreamRecorder).newSubstream();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/RetriableStreamTest.java]: Use `doReturn(mockStream1, mockStream2, mockStream3).when(retriableStreamRecorder).newSubstream()` instead?

Comment (dapengzhang0) [reply core/src/test/java/io/grpc/internal/RetriableStreamTest.java]: Seems the current version mockito we used does not support vargs.

File: core/src/main/java/io/grpc/internal/RetriableStream.java [hunk 4]
```diff
@@ -166,46 +176,48 @@ public final void start(ClientStreamListener listener) {
...
+    Substream substream = newSubstream();
     drain(substream);
 
     // TODO(zdapeng): schedule hedging if needed
   }
 
   @Override
   public final void cancel(Status reason) {
-    if (commit0(new NoopClientStream())) {
+    Substream noopSubstream = newSubstream();
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/RetriableStream.java]: Uhh, good catch! The name is so errorprone.

File: core/src/test/java/io/grpc/internal/RetriableStreamTest.java (hunk 2 above)
Comment (dapengzhang0) [reply core/src/test/java/io/grpc/internal/RetriableStreamTest.java]: I wouldn't do this if mockito doReturn supported varargs. And I don't like `when(a.method()).thenReturn(b).thenReturn(c)`

File: core/src/main/java/io/grpc/internal/RetriableStream.java [hunk 5]
```diff
@@ -536,6 +548,21 @@ State drained(Substream substream) {
       return new State(buffer, drainedSubstreams, winningSubstream, cancelled, passThrough);
     }
 
+    /** The given substream is closed. */
+    @CheckReturnValue
+    @GuardedBy("lock")
+    State closed(Substream substream) {
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/RetriableStream.java]: Done.

File: core/src/main/java/io/grpc/internal/RetriableStream.java (hunk 1 above)
Comment (dapengzhang0) [reply core/src/main/java/io/grpc/internal/RetriableStream.java]: Done.

File: core/src/main/java/io/grpc/internal/RetriableStream.java [hunk 6]
```diff
@@ -550,7 +550,7 @@ State drained(Substream substream) {
     /** The given substream is closed. */
     @CheckReturnValue
     @GuardedBy("lock")
-    State closed(Substream substream) {
+    State sbustreamClosed(Substream substream) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/RetriableStream.java]: s/sbustream/substream/

File: core/src/main/java/io/grpc/internal/RetriableStream.java [hunk 7]
```diff
@@ -505,32 +526,49 @@ State cancelled() {
...
         buffer = null;
       }
 
       return new State(buffer, drainedSubstreams, winningSubstream, cancelled, passThrough);
     }
 
+    /** The given substream is closed. */
+    @CheckReturnValue
+    @GuardedBy("lock")
+    State sbustreamClosed(Substream substream) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/internal/RetriableStream.java]: s/sbustream/substream/

(since github thinks my other comment is resolved)

**Pull request merged**